compilers/concrete-compiler/compiler/include/concretelang/Transforms/TypeInferenceRewriter.h

```diff
@@ -0,0 +1,327 @@
+// Exceptions. See
+// https://github.com/zama-ai/concrete-compiler-internal/blob/main/LICENSE.txt
+// for license information.
+
+// Infrastructure for rewriting IR after running type inference using
+// `ForwardTypeInferenceAnalysis` and `BackwardTypeInferenceAnalysis`
+//
+// The `TypeInferenceRewriter` class can be used as a base for a
+// rewriter applying the types inferred through type inference. Final
+// types are inferred by retrieving the inferred types for the
+// respective values from the data flow solver and by invoking the type
+// resolver one last time.
+//
+// Conflicts can be handled by overriding the `handleConflict`
+// method. By default, this method generates a
+// `TypeInference.unresolved_conflict` operation.
+
+#ifndef CONCRETELANG_TRANSFORMS_TYPEINFERENCEREWRITER_H
+#define CONCRETELANG_TRANSFORMS_TYPEINFERENCEREWRITER_H
+
+#include <concretelang/Analysis/TypeInferenceAnalysis.h>
+#include <concretelang/Dialect/TypeInference/IR/TypeInferenceOps.h>
+
+#include <mlir/Analysis/DataFlow/SparseAnalysis.h>
+#include <mlir/Dialect/Func/IR/FuncOps.h>
+#include <mlir/IR/IRMapping.h>
+#include <mlir/IR/PatternMatch.h>
+
+namespace mlir {
+namespace concretelang {
+
+// Rewrites the types in a module based on the results contained in
+// the states of a `DataFlow` solver that has run type analysis.
+class TypeInferenceRewriter {
+public:
+  TypeInferenceRewriter(const mlir::DataFlowSolver &solver,
+                        TypeResolver &typeResolver)
+      : solver(solver), typeResolver(typeResolver) {}
+
+  virtual ~TypeInferenceRewriter() {}
+
+  // Rewrites a module. If rewriting fails, e.g., due to unresolved
+  // types, an error is emitted for the corresponding op, the module
+  // is left partially rewritten and the function returns
+  // `mlir::failure()`.
+  mlir::LogicalResult rewrite(mlir::ModuleOp module) {
+    mlir::IRRewriter rewriter(module.getContext());
+
+    if (module
+            .walk([&](mlir::func::FuncOp func) {
+              if (rewrite(func, rewriter).failed()) {
+                return mlir::WalkResult::interrupt();
+              } else {
+                return mlir::WalkResult::advance();
+              }
+            })
+            .wasInterrupted()) {
+      return mlir::failure();
+    }
+
+    this->postRewriteHook(module);
+
+    return mlir::success();
+  }
+
+protected:
+  // Rewrite a single function
+  mlir::LogicalResult rewrite(mlir::func::FuncOp func,
+                              mlir::IRRewriter &rewriter) {
+    rewriter.setInsertionPointAfter(func);
+
+    // Gather inferred types an run the solver a last time on the
+    // function to get the actual types for the arguments and return
+    // values
+    LocalInferenceState inferredTypes =
+        TypeInferenceUtils::getLocalInferenceState(solver, func);
+    LocalInferenceState resolvedTypes;
+    typeResolver.resolve(func, resolvedTypes, inferredTypes);
```

Comment:
So if I understand correctly, our inference only made the inference inside of the function block, and here would like types from the inside to be propagated to the function signature basically. I don't understand why the type resolver is needed. I would expect the dataflow analysis to do that. What am I missing ?

Reply:
The issue is with resolved types, which are not explicitly stored in lattice values corresponding to the function arguments. Without the final invocation of the type resolver, the types of the arguments would appear as missing.